core/commands/dag/dag.go

```diff
@@ -241,3 +277,487 @@ var DagResolveCmd = &cmds.Command{
 	},
 	Type: ResolveOutput{},
 }
+
+// It is not guaranteed that a root in a header is actually present in the same ( or any )
+// .car file. This is the case in version 1, and ideally in further versions too
+// Accumulate any root CID seen in a header, and supplement its actual node if/when encountered
+// We will attempt a pin *only* at the end in case all car files were well formed
+//
+// The boolean value indicates whether we have encountered the root within the car file's
+type expectedRootsSeen map[cid.Cid]bool
+type importResult struct {
+	roots *expectedRootsSeen
+	err   error
+}
+
+var DagImportCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Import the contents of .car files",
+		ShortDescription: `
+'ipfs dag import' parses .car files and adds all objects to the blockstore.
+By default, after all car files have been processed, an attempt is made to
+pin each individual root specified in the car headers, before GC runs again.
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.FileArg("path", true, true, "The path of a .car file.").EnableStdin(),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption(progressOptionName, "p", "Display progress on CLI. Defaults to true when STDERR is a TTY"),
+		cmds.BoolOption(silentOptionName, "No output."),
+		cmds.BoolOption(pinRootsOptionName, "Pin optional roots listed in the .car headers after importing.").WithDefault(true),
+	},
+	PreRun: func(req *cmds.Request, env cmds.Environment) error {
+		silent, _ := req.Options[silentOptionName].(bool)
+		encType, _ := req.Options[cmds.EncLong].(string)
+
+		if silent || encType != "text" {
+			// force-disable progress unless on non-silent CLI
+			req.Options[progressOptionName] = false
```

Comment:
If I ask for progress but set the encoding type to `json`, I'd expect json progress updates. I wouldn't mess with this in pre-run.